scripts/setup_ubuntu.sh

```diff
@@ -21,14 +21,12 @@ for arg in "$@"; do
 done
 
 run_as_original_user() {
-  # The second "$@" is expanded in this script and becomes the arguments to a
-  # bash script `"$@"` that simply executes its command-line arguments. This
-  # protects characters like < from being interpreted as redirections if $@
-  # were expanded directly.
+  # $* joins all arguments into a single string.
+
   if [[ -n "${ORIGINAL_USER}" ]]; then
-    su - "${ORIGINAL_USER}" bash -c '"$@"' -- "$@"
+    su - "${ORIGINAL_USER}" bash -c "$*"
```

Comment:
The code was written to support things like:
```
run_as_original_user foo pip install --upgrade pip<=20.3.4
```
as strictly stringifying would lead to `<` being interpreted as a pipe redirect.

Reply:
Maybe a cleaner solution here is to make a `requirements.txt`? Feels weird to have extra complexity we're not using here.

Reply:
We could generate and use a requirements file in the script. I don't think we could include the requirement file in the repo because in at least one script the packages are selected based on the environment the script is executing in.

That said let's not make this `run_as_original_user` less safe. If it's failing let's fix it right. Bash is terrible, and not quoting things properly will just make things more confusing. When and how was `run_as_original_user` failing? It worked for me and I tested it a good amount, so it doesn't universally fail like your comment implies.

Reply:
actually it seems like the `$*` approach already handles the redirect stuff correctly. I still think a `requirements.txt` would be easier to deal with and would avoid this confusion entirely

Reply:
maybe it has to do with the fact the I'm using a weird shell and `su` executes the command in your login shell? I'm using `xonsh` as my login shell, so maybe the difference in variable expansion is somehow incompatible there.

Reply:
I'm 99% sure it doesn't. This is what I get for a simple test:

```
$ function tttttt() { bash -c "$*"; }
$ tttttt echo "test<test"
bash: test: No such file or directory
```
(The `<` is being interpreted)

```
$ function tttttt() { bash -c '"$@"' -- "$@"; }
$ tttttt echo "test<test"
test<test
```
(The `<` is being echoed)

Yes. I think generating a requirements files (which we should NOT name `requirements.txt` because it's not actually package requirements in the standard sense) is great. But we should still keep the quote correct version of this function unless it can be removed entirely (which I don't think it can be).

Reply:
> `xonsh` as my login shell

That could be part of the issue. But since the script is run with `/bin/bash` and the `su` command immediately calls `bash` I don't see how `xonsh` could be involved. At most it should be passing through the arguments without interpretation. But it's possible there is some login shell step I'm not aware of in `su`.

Reply:
ah that's weird, I tested it by making the setup script add a constraint on the package version and that seemed to work. If we do switch to generating the requirements file can't we just change this function to `install_pip_requirements_as_user` instead and then not worry about the redirection stuff? There's no uses of this function to do anything besides call `pip install`. 

Hmm, yeah the parameter expansion should be done by bash before the `su` call so my login shell shouldn't matter...

Reply:
Yes that's true.

As a side note, I think you should probably debug the shell issue in your environment. If your environment is misinterpreting shell scripts you could have some really hard bugs down the road. And it seems unlikely that this is the only case where scripts will execute incorrectly because this is not a very complex set of substitutions semantically.

Reply:
ok, it was my login shell. running `su -s $(which bash) - ....` worked. I'll just add the `-s` flag here for now.

Reply:
Perfect. Thanks for debugging this. And I'll try to remember this. Using `su` in a script isn't common, but it does happen and this issue could also affect people using `zsh`.

I'm still open to switching to use a file. But it's probably not worth that much engineering time.